packages/jazz-tools/src/schema/coMap.ts

```diff
@@ -0,0 +1,364 @@
+    const schemaDef = this as unknown as Record<
+      string,
+      { [SchemaInit]: Schema }
+    >;
+
+    return (schemaDef[key as string]! || schemaDef[ItemsSym])?.[
+      SchemaInit
+    ] as Schema;
+  }
+
+  /**
+   * Create a new CoMap with the given initial values and owner.
+   *
+   * The owner (a Group or Account) determines access rights to the CoMap.
+   *
+   * The CoMap will immediately be persisted and synced to connected peers.
+   *
+   * @example
+   * ```ts
+   * const person = Person.create({
+   *   name: "Alice",
+   *   age: 42,
+   *   pet: cat,
+   * }, { owner: friendGroup });
+   * ```
+   *
+   * @category Creation
+   **/
+  static create<M extends CoMapSchema>(
+    this: CoMapSchemaClass<M>,
+    init: Simplify<CoMapInit<M>>,
+    options?:
+      | {
+          owner: Account | Group;
+          unique?: CoValueUniqueness["uniqueness"];
+        }
+      | Account
+      | Group,
+  ) {
+    const schema = new this();
+
+    const { owner, uniqueness } = parseCoValueCreateOptions(options);
+    return CoMapInstance.fromInit(schema, init, owner, uniqueness);
+  }
+
+  /**
+   * Declare a Record-like CoMap schema, by extending `CoMap.Record(...)` and passing the value schema using `co`. Keys are always `string`.
+   *
+   * @example
+   * ```ts
+   * import { co, CoMap } from "jazz-tools";
+   *
+   * class ColorToFruitMap extends CoMap.Record(
+   *  co.ref(Fruit)
+   * ) {}
+   *
+   * // assume we have map: ColorToFruitMap
+   * // and strawberry: Fruit
+   * map["red"] = strawberry;
+   * ```
+   *
+   * @category Declaration
+   */
+  static Record<Value>(value: IfCo<Value, Value>) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+    class RecordLikeCoMap extends CoMapSchema {
+      [ItemsSym] = value;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+    interface RecordLikeCoMap extends Record<string, Value> {}
+
+    return RecordLikeCoMap;
+  }
+}
+
+export type CoKeys<Map extends object> = Exclude<
+  keyof Map & string,
+  keyof CoMapSchema
+>;
+
+type ForceRequiredRef<V> = V extends co<InstanceType<
+  CoMapSchemaClass<any>
+> | null>
+  ? NonNullable<V>
+  : V extends co<CoMapSchema | undefined>
+    ? CoMapInstance<NonNullable<UnCo<V>>> | null
+    : V;
+
+export type CoMapInit<Map extends object> = {
+  [Key in CoKeys<Map> as undefined extends Map[Key]
+    ? never
+    : IfCo<Map[Key], Key>]: ForceRequiredRef<Map[Key]>;
+} & {
+  [Key in CoKeys<Map> as IfCo<Map[Key], Key>]?: ForceRequiredRef<Map[Key]>;
+};
+
+export class CoMapInstance<M extends CoMapSchema> extends CoValuePrototype {
+  declare $raw: RawCoMap;
+  declare $schema: M;
+  declare $id: ID<M>;
+  declare $resolutionNode: CoValueResolutionNode<M> | undefined;
+  refs = new Map<string, CoMapInstance<any>>();
+  private $lastUpdateTx: number;
+
+  constructor(
+    schema: M,
+    raw: RawCoMap,
+    resolutionNode?: CoValueResolutionNode<M>,
```

Comment:
Passing the "ResolutionNode" to be able to increase the scope of the subscription.

More below.